zend_abstract_interface/interceptor/php7/interceptor.c

```diff
@@ -0,0 +1,1108 @@
+        (zend_string *)&zend_pass_function_name,     /* name              */
+        NULL,                   /* scope             */
+        NULL,                   /* prototype         */
+        0,                      /* num_args          */
+        0,                      /* required_num_args */
+        NULL,                   /* arg_info          */
+        ZEND_FN(pass),          /* handler           */
+        NULL,                   /* module            */
+        {0}                     /* reserved          */
+};
+#endif
+
+typedef struct {
+    zai_hook_memory_t hook_data;
+    zend_execute_data *execute_data;
+    bool implicit;
+} zai_interceptor_frame_memory;
+
+typedef struct {
+    zai_interceptor_frame_memory frame;
+    const zend_op *return_op;
+    zend_op resumption_ops[3];
+    bool resumed;
+    uint32_t temporary;
+} zai_interceptor_generator_frame_memory;
+
+__thread HashTable zai_hook_memory;
+// execute_data is 16 byte aligned (except when it isn't, but it doesn't matter as zend_execute_data is big enough
+// our goal is to reduce conflicts
+static inline void zai_hook_memory_table_insert(zend_execute_data *index, zai_interceptor_frame_memory *inserting) {
+    zend_hash_index_update_mem(&zai_hook_memory, ((zend_ulong)index) >> 4, inserting, sizeof(*inserting));
+}
+
+static inline void *zai_hook_memory_table_insert_generator(zend_execute_data *index, zai_interceptor_generator_frame_memory *inserting) {
+    return zend_hash_index_update_mem(&zai_hook_memory, ((zend_ulong)index) >> 4, inserting, sizeof(*inserting));
+}
+
+static inline bool zai_hook_memory_table_find(zend_execute_data *index, zai_interceptor_frame_memory **found) {
+    return zai_hook_table_find(&zai_hook_memory, ((zend_ulong)index) >> 4, (void **)found);
+}
+
+static inline bool zai_hook_memory_table_del(zend_execute_data *index) {
+    return zend_hash_index_del(&zai_hook_memory, ((zend_ulong)index) >> 4);
+}
+
+#define ZAI_INTERCEPTOR_CUSTOM_EXT 0xda8ad065
+
+static void zai_set_ext_nop(zend_op *op) {
+    memset(op, 0, sizeof(zend_op));
+    op->lineno = CG(zend_lineno);
+    SET_UNUSED(op->result);
+    SET_UNUSED(op->op1);
+    SET_UNUSED(op->op2);
+    op->opcode = ZEND_EXT_NOP;
+    op->extended_value = ZAI_INTERCEPTOR_CUSTOM_EXT;
+}
+
+void zai_interceptor_op_array_ctor(zend_op_array *op_array) {
+    // push our own EXT_NOP onto the op_array start
+    zend_op *op = &op_array->opcodes[op_array->last++];
+    zai_set_ext_nop(op);
+    // EXT_STMT is skipped by compiler when determining "very first" instructions
+    op->opcode = ZEND_EXT_STMT;
+
+    // For reasons this is not set on eval()'ed code, leading to pass_two handler not being called
+    CG(compiler_options) |= ZEND_COMPILE_HANDLE_OP_ARRAY;
+}
+
+static inline bool zai_is_func_recv_opcode(zend_uchar opcode) {
+    return opcode == ZEND_RECV || opcode == ZEND_RECV_INIT || opcode == ZEND_RECV_VARIADIC;
+}
+
+void zai_interceptor_op_array_pass_two(zend_op_array *op_array) {
```

Comment:
I'm not very familiar with `pass_two`. What's the objective of us using this hook?

Reply:
the pass_two handler is invoked immediately after first pass complication has finished. I.e. it's _the_ chance to do changes to an op_array before pass_two optimizes some things, resolves labels etc.

Reply:
Thanks. What are we using it for? I don't understand the nuts and bolts of  `ZAI_INTERCEPTOR_CUSTOM_EXT` and such. Is there a high level comment about this somewhere else already? If not can we add one here or elsewhere and refer to it?

In general I don't understand the details of the opcodes well enough to say whether anything is right, and that's okay but some breadcrumbs for newcomers like me would be helpful.